Update to comply fully with Rev. 3.0 of OmniLink

ecdye opened the pull request:
Added all ExtendedObjectStatuses (mostly for fun) to ensure full code
coverage of OmniLink features. Fixed several bugs that I found along the
way, including a missing Thermostat Properties value that indicates the
capabilities of a given thermostat.

Rewrote all documentation to preserve the inner working of the OmniLink
protocol in a nice manner for the future developers of this project.
Additionally, as a way to manually format the documentation the code
style of the project has also been updated and unified to one standard.

Signed-off-by: Ethan Dye <mrtops03@gmail.com>

randye007 commented:
@ecdye  - Thank-you! Thank-you! Thank-you! If you need someone to test your jar file, please let me know. I'd be happy to test with you.

digitaldan commented:
Hi , thanks for the PR.  There's a lot of changes combined in this PR, so it may take a bit to review.  In the mean time i ran the example from the README based on this branch and am seeing an error. Can you please take a look?  Thanks.

```
[main] ERROR com.digitaldan.jomnilinkII.OmniInvalidResponseException - Invalid message type received: NegativeAcknowledge()
[main] ERROR com.digitaldan.jomnilinkII.examples.Main - Invalid Response
com.digitaldan.jomnilinkII.OmniInvalidResponseException
	at com.digitaldan.jomnilinkII.Connection.reqObjectStatus(Connection.java:601)
	at com.digitaldan.jomnilinkII.examples.Main.main(Main.java:233)
```

ecdye commented:
Hmm, i had issues with the Extended thermostat status before as well, @digitaldan do you run any omnistats? what model is your omni?

digitaldan commented on src/main/java/com/digitaldan/jomnilinkII/examples/Main.java:
```diff
@@ -28,240 +28,319 @@
+            for (int i = 0; i < thermos.length; i++) {
+                logger.info(thermos[i].toString());
+            }
+            status = c.reqObjectStatus(Message.OBJ_TYPE_THERMO, 1, 16, true); // It appears that the maximum number of
```
is this a testing line ?  its throwing an exception on my system since i don't have thermostats on my omni

digitaldan commented:
I'm using an omni LTE with no thermostats for testing,  but see my comment on the code about why the error is being thrown.

ecdye commented:
Ahh, I see the issue, I just need to put a cap on the max number if the capacities exceeds 16 I'll work on a fix.

digitaldan commented:
>I just need to put a cap on the max number if the capacities exceeds 16 

I don't think thats the issue.  The way the protocol works is that you have to first query the number of thermostats (or any object capacity) like so:

`int max_thermos = c.reqObjectTypeCapacities(Message.OBJ_TYPE_THERMO).getCapacity();`

Then you cycle through requesting the status for each one, this is per the omnilink spec.   The original code above this line does the correct query, so i'm not sure why this line was added.

`status = c.reqObjectStatus(Message.OBJ_TYPE_THERMO, 1, max_thermos);`

digitaldan commented:
So i'm not sure how to review this.  every file (130) has been touched and reformatted so i have no idea what functionality was added, fixed or removed.  Generally you never want to mix up changes like this.  It would be nice if you applied formatting first as a separate PR, then submit your fixes as a follow up PR so we can track changes.  Also it would be nice to know what IDE or tool you are using to auto format the source files.

ecdye commented:
> So i'm not sure how to review this. every file (130) has been touched and reformatted so i have no idea what functionality was added, fixed or removed. Generally you never want to mix up changes like this. It would be nice if you applied formatting first as a separate PR, then submit your fixes as a follow up PR so we can track changes. Also it would be nice to know what IDE or tool you are using to auto format the source files.

So I was using spotless to format the files with similar settings to openhab. I can try to separate the changes if you would like.

>Then you cycle through requesting the status for each one, this is per the omnilink spec. The original code above this line does the correct query, so i'm not sure why this line was added.

It was to allow for testing of the newer extend object status functionality, it should now work correctly, and only request the status from setup units.

If it helps, most of the code changes were made to the files related to ObjectStatus, MessageFactory.java, and Connection.java

digitaldan commented:
@ecdye I do appreciate you contributing to the library, i'm glad its of use and still valuable.  This commit however has now grown very large and is not really acceptable for review.   I think spotless is a good thing to use here, but if we do that, we should add it as a maven plugin so its part of the project and can be applied in a standard way.  I have a PR open, #34 which adds the spotless plugin, a license header file and applies spotless to the source.  I would like to merge this first, then have you add your documentation changes and any functionality changes as another pr based on this.  If you want to combine the later into a single PR that would be ok, although smaller PRs that have a single responsibility are greatly appreciated. 

wdyt?

ecdye commented:
Sounds fair enough, the hard work of redocumenting is done, its relatively speaking easy to restructure the PR from here.

ecdye committed:
Update to comply fully with Rev. 3.0 of OmniLink

Added all ExtendedObjectStatuses (mostly for fun) to ensure full code
coverage of OmniLink features. Fixed several bugs that I found along the
way, including a missing Thermostat Properties value that indicates the
capabilities of a given thermostat.

Rewrote all documentation to preserve the inner workings of the OmniLink
protocol in a nice manner for the future developers of this project.

Signed-off-by: Ethan Dye <mrtops03@gmail.com>

ecdye commented:
@digitaldan if you find #35 acceptable, if merged I could simply force push an updated commit that was based off of it and allow you to easily continue the review process without all of the formatting changes on this PR.

digitaldan commented:
Yes, that sounds good, thank you. I just merged #35, so feel free to re push to this branch.  Thanks!

ecdye commented:
> Yes, that sounds good, thank you. I just merged #35, so feel free to re push to this branch. Thanks!

Done, should be much easier to review now.

digitaldan commented on src/main/java/com/digitaldan/jomnilinkII/OmniInvalidResponseException.java:
```diff
@@ -9,13 +9,18 @@
 	public OmniInvalidResponseException(Message response) {
 		super();
 		this.response = response;
+		logger.error("Invalid message type received: {}", response.toString());
```
Since this is an exception class, can you remove the logging from it? Who ever gets this should do the logging.  Also i avoid logging anything other then trace/debug/warn since this is a shared library and I don't want to pollute the calling application with logging under normal conditions.

ecdye replied:
Yeah, that was supposed to be removed. I had added it to help myself figure out what was going on but forgot to remove it.

ecdye committed:
Remove logging from OmniInvalidResponseException

Signed-off-by: Ethan Dye <mrtops03@gmail.com>

digitaldan commented on src/main/java/com/digitaldan/jomnilinkII/examples/Main.java:
```diff
@@ -123,121 +121,259 @@ public void notConnectedEvent(Exception e) {
-			for (int i = 0; i < auxs.length; i++) {
-				logger.info(auxs[i].toString());
+			if (expNum > 0) {
+				ObjectStatus status = c.reqObjectStatus(Message.OBJ_TYPE_EXP, 1, expNum);
```
So this is throwing an exception since i don't have an expansion and `expNum` is hardcoded to "8" above in the file, `consoleNum` is also hardcoded

ecdye replied:
That is because they both don't return valid responses for an object properties request. Any ideas?

ecdye replied:
@digitaldan I commented out the code but left it there as an example, is that acceptable?

ecdye committed:
Comment out potentially broken code

Signed-off-by: Ethan Dye <mrtops03@gmail.com>

ecdye committed:
Fix typo

Signed-off-by: Ethan Dye <mrtops03@gmail.com>

digitaldan commented:
Thanks, sorry for the delay.  So the querying for number of exp enclosures does work on my system, but it returns 1, while it should return 0, so rather than chase that down now, we can punt for a later time.
